File: practical-11-product-cipher/product_cipher.py
import string


def caesar_encrypt(text, shift):
    """Apply Caesar cipher encryption."""
    result = ""
    for char in text:
        if char.isupper():
            result += chr((ord(char) - 65 + shift) % 26 + 65)
        elif char.islower():
            result += chr((ord(char) - 97 + shift) % 26 + 97)
        else:
            result += char
    return result


def caesar_decrypt(text, shift):
    """Apply Caesar cipher decryption."""
    return caesar_encrypt(text, 26 - shift)


def get_key_order(keyword):
    """Get column order based on alphabetical sorting of keyword."""
    keyword = keyword.upper().replace(" ", "")
    indexed = [(char, i) for i, char in enumerate(keyword)]
    sorted_indexed = sorted(indexed, key=lambda x: (x[0], x[1]))

    order = [0] * len(keyword)
    for new_pos, (_, old_pos) in enumerate(sorted_indexed):
        order[old_pos] = new_pos
    return order


def row_transposition_encrypt(text, keyword):
    """Apply Row Transposition encryption."""
    keyword = keyword.upper().replace(" ", "")
    key_len = len(keyword)

    padding_needed = (key_len - len(text) % key_len) % key_len
    text = text + "X" * padding_needed

    rows = []
    for i in range(0, len(text), key_len):
        rows.append(list(text[i : i + key_len]))

    order = get_key_order(keyword)
    encrypted = ""

    for col_idx in range(key_len):
        original_col = order.index(col_idx)
        for row in rows:
            if col_idx < len(row):
                encrypted += row[original_col]

    return encrypted


def row_transposition_decrypt(text, keyword):
    """Apply Row Transposition decryption."""
    keyword = keyword.upper().replace(" ", "")
    key_len = len(keyword)
    num_rows = (len(text) + key_len - 1) // key_len

    order = get_key_order(keyword)
    grid = [[""] * key_len for _ in range(num_rows)]

    char_idx = 0
    sorted_cols = sorted(range(key_len), key=lambda x: (keyword[x], x))

    for new_col, original_col in enumerate(sorted_cols):
        for row in range(num_rows):
            if char_idx < len(text):
                grid[row][original_col] = text[char_idx]
                char_idx += 1

    plaintext = ""
    for row in grid:
        plaintext += "".join(row)

    return plaintext.rstrip("X")


def product_encrypt(plaintext, caesar_shift, transposition_key):
    """Encrypt using Product cipher (Caesar + Row Transposition)."""
    step1 = caesar_encrypt(plaintext, caesar_shift)
    step2 = row_transposition_encrypt(step1, transposition_key)
    return step2


def product_decrypt(ciphertext, caesar_shift, transposition_key):
    """Decrypt Product cipher (reverse: Row Transposition + Caesar)."""
    step1 = row_transposition_decrypt(ciphertext, transposition_key)
    step2 = caesar_decrypt(step1, caesar_shift)
    return step2


if __name__ == "__main__":
    text = "SECRET MESSAGE"
    shift = 3
    key = "PYTHON"

    encrypted = product_encrypt(text, shift, key)
    decrypted = product_decrypt(encrypted, shift, key)

    print(f"Original: {text}")
    print(f"Caesar Shift: {shift}")
    print(f"Transposition Key: {key}")
    print(f"Encrypted: {encrypted}")
    print(f"Decrypted: {decrypted}")

File: c programs/product_cipher.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

void caesar_encrypt(const char *text, int shift, char *result) {
    for (int i = 0; text[i] != '\0'; i++) {
        if (isupper(text[i])) {
            result[i] = (char)((text[i] - 65 + shift) % 26 + 65);
        } else if (islower(text[i])) {
            result[i] = (char)((text[i] - 97 + shift) % 26 + 97);
        } else {
            result[i] = text[i];
        }
    }
    result[ strlen(text) ] = '\0';
}

void caesar_decrypt(const char *text, int shift, char *result) {
    caesar_encrypt(text, 26 - shift, result);
}

void get_key_order(const char *keyword, int order[]) {
    int key_len = strlen(keyword);
    
    for (int i = 0; i < key_len; i++) {
        order[i] = i;
    }
    
    for (int i = 0; i < key_len - 1; i++) {
        for (int j = i + 1; j < key_len; j++) {
            char ci = toupper(keyword[order[i]]);
            char cj = toupper(keyword[order[j]]);
            
            if (ci > cj || (ci == cj && order[i] > order[j])) {
                int temp = order[i];
                order[i] = order[j];
                order[j] = temp;
            }
        }
    }
}

void row_transposition_encrypt(const char *text, const char *keyword, char *result) {
    int key_len = strlen(keyword);
    
    int padded_len = ((strlen(text) + key_len - 1) / key_len) * key_len;
    char buf[100];
    
    for (int i = 0; i < strlen(text); i++) {
        buf[i] = text[i];
    }
    for (int i = strlen(text); i < padded_len; i++) {
        buf[i] = 'X';
    }
    buf[padded_len] = '\0';
    
    int num_rows = padded_len / key_len;
    char grid[10][20];
    
    for (int i = 0; i < num_rows; i++) {
        for (int j = 0; j < key_len; j++) {
            grid[i][j] = buf[i * key_len + j];
        }
    }
    
    int order[20];
    get_key_order(keyword, order);
    
    int r_idx = 0;
    for (int col = 0; col < key_len; col++) {
        int orig_col = -1;
        for (int i = 0; i < key_len; i++) {
            if (order[i] == col) {
                orig_col = i;
                break;
            }
        }
        
        for (int row = 0; row < num_rows; row++) {
            result[r_idx] = grid[row][orig_col];
            r_idx++;
        }
    }
    result[r_idx] = '\0';
}

void row_transposition_decrypt(const char *text, const char *keyword, char *result) {
    int key_len = strlen(keyword);
    int text_len = strlen(text);
    int num_rows = (text_len + key_len - 1) / key_len;
    
    int order[20];
    get_key_order(keyword, order);
    
    char grid[10][20];
    for (int i = 0; i < num_rows; i++) {
        for (int j = 0; j < key_len; j++) {
            grid[i][j] = '\0';
        }
    }
    
    int char_idx = 0;
    for (int new_col = 0; new_col < key_len; new_col++) {
        int orig_col = order[new_col];
        for (int row = 0; row < num_rows; row++) {
            if (char_idx < text_len) {
                grid[row][orig_col] = text[char_idx];
                char_idx++;
            }
        }
    }
    
    int r_idx = 0;
    for (int i = 0; i < num_rows; i++) {
        for (int j = 0; j < key_len && grid[i][j] != '\0'; j++) {
            result[r_idx] = grid[i][j];
            r_idx++;
        }
    }
    
    while (r_idx > 0 && result[r_idx - 1] == 'X') {
        r_idx--;
    }
    result[r_idx] = '\0';
}

void product_encrypt(const char *plaintext, int caesar_shift, const char *transposition_key, char *result) {
    char step1[100];
    caesar_encrypt(plaintext, caesar_shift, step1);
    row_transposition_encrypt(step1, transposition_key, result);
}

void product_decrypt(const char *ciphertext, int caesar_shift, const char *transposition_key, char *result) {
    char step1[100];
    row_transposition_decrypt(ciphertext, transposition_key, step1);
    caesar_decrypt(step1, caesar_shift, result);
}

int main() {
    const char *text = "SECRET MESSAGE";
    int shift = 3;
    const char *key = "PYTHON";
    
    char encrypted[100];
    char decrypted[100];
    
    product_encrypt(text, shift, key, encrypted);
    product_decrypt(encrypted, shift, key, decrypted);
    
    printf("Original: %s\n", text);
    printf("Caesar Shift: %d\n", shift);
    printf("Transposition Key: %s\n", key);
    printf("Encrypted: %s\n", encrypted);
    printf("Decrypted: %s\n", decrypted);
    
    return 0;
}
